### PR TITLE
[Docs] [AutoDiff] Add default derivative subsection and fix minor issues.

### PR DESCRIPTION
**Pull request by rxwei**
* Add section "Default derivatives", which documents what's discussed [here](https://forums.swift.org/t/differentiable-programming-mega-proposal/28547/118).
* `.withoutDerivative()` -> `withoutDerivative(at:)`
* Fix syntax error in default derivatives.

### CODE REVIEW

File: docs/DifferentiableProgramming.md [hunk 1]
```diff
@@ -1750,6 +1753,73 @@ where Self: Differentiable & FloatingPoint, Self == Self.TangentVector {
...
+
+extension P {
+    @derivative(of: foo(x:))
+    func _(_ x: Float) -> (value: Float, differential: (Float) -> Float) {
+        (value: foo(x), differential: { _ in 42 })
+    }
+}
+
+struct S: P {
+    @differentiable(default) // Inherits from `P.foo(_:)`.
```
Comment (dan-zheng) [inline docs/DifferentiableProgramming.md]: How about adding some more detail:
```suggestion
    @differentiable(default) // Inherit default derivative for `P.foo(_:)`.
```

**Review by dan-zheng — APPROVED** [review]
LGTM! Though I'd recommend waiting for @marcrasi to review also.

File: docs/DifferentiableProgramming.md (hunk 1 above)
Comment (rxwei) [reply docs/DifferentiableProgramming.md]: Done.

File: docs/DifferentiableProgramming.md [hunk 2]
```diff
@@ -1750,6 +1753,73 @@ where Self: Differentiable & FloatingPoint, Self == Self.TangentVector {
...
 
+#### Default derivatives
+
+In a protocol extension, class definition, or class extension, providing a
+derivative or transpose for a protocol extension or a non-final class member is
+considered as providing a default derivative for that member. Types that conform
+to the protocol or inherit from the class can inherit the default derivative.
+
+If the original member does not have a `@differentiable` attribute, a default
+derivative is implicitly add to all conforming/overriding implementations.
```
Comment (bartchr808) [inline docs/DifferentiableProgramming.md]:
```suggestion
derivative is implicitly added to all conforming/overriding implementations.
```

File: docs/DifferentiableProgramming.md [hunk 3]
```diff
@@ -1750,6 +1753,73 @@ where Self: Differentiable & FloatingPoint, Self == Self.TangentVector {
...
+let s = S()
+let d = derivative(at: 0) { x in
+   s.foo(x)
+} // ==> 42
+```
+
+When a protocol requirement or class member is marked with `@differentiable`, it
+is considered as a _differentiability customization point_. This means that all
+conforming/overriding implementation must provide a corresponding
+`@differentiable` attribute, which causes the implementation to be will be
```
Comment (bartchr808) [inline docs/DifferentiableProgramming.md]: I'm guessing it should be "to be" based on how it reads:
```suggestion
`@differentiable` attribute, which causes the implementation to be
```

**Review by bartchr808 — APPROVED** [review]

**Review by marcrasi — APPROVED** [review]
Looks like a good solution to me.

I have one question. What happens if a user defines a `@derivative` of a `@differentiable(default)` method?

**Comment by dan-zheng** [comment]
> I have one question. What happens if a user defines a `@derivative` of a `@differentiable(default)` method?

It seems to me that should be rejected: `@derivative` and `@differentiable(default)` are in conflict.

**Comment by rxwei** [comment]
> Looks like a good solution to me.
> 
> I have one question. What happens if a user defines a `@derivative` of a `@differentiable(default)` method?

 I think this should be banned if it’s in the same module.  Cross module should be allowed for flexibility, but can be banned in the initial proposal.

**Comment by rxwei** [comment]
@swift-ci please smoke test and merge

File: docs/DifferentiableProgramming.md (hunk 3 above)
Comment (rxwei) [reply docs/DifferentiableProgramming.md]: Good catch!

**Comment by bartchr808** [comment]
This:

> I have one question. What happens if a user defines a @derivative of a @differentiable(default) method?

Made me think about whether there would be any problems with dynamic lookup of functions, but as far as I can tell there aren't any weird cases.

**Comment by rxwei** [comment]
That should be orthogonal, as statically the only thing involved is `subscript(dynamicMember:)`. Derivative registration only cares about member declarations in terms of static lookup.

**Pull request merged**